src/interfaces/particle_distribution.jl

```diff
@@ -0,0 +1,244 @@
+* [`_weight(s::ParticleSampleable,x)`](@ref): return the weight of a given sample `x`
+* [`is_exact(s::ParticleSampleable)`](@ref): return wether or not a particle-sampleable is exact
+
+Optionally, one can enhance the calculation of weights by providing
+
+* [`_assert_valid_input_type(s::ParticleSampleable,x)`](@ref): assert input has the correct type
+* [`_assert_valid_input(s::ParticleSampleable,x)`](@ref): assert input has correct properties
+* [`_post_processing(s::ParticleSampleable,x,res)`](@ref): apply some postprocesing to the result of `_weight`.
+
+See [`weight`](@ref) for details. Furthermore, one can provide custom four-momentum types used for
+the generation of samples by implementing
+
+* [`_momentum_type(s::ParticleSampleable)`](@ref): return the momentum type used
+
+The actual sampling function depends on the variate form:
+
+!!! note "Single particle sampler"
+    For `SingleParticleVariate`samplers, one needs to implement the single sample version `rand`:
+
+    ```julia
+    Distributions.rand(
+        rng::Random.AbstractRNG,
+        s::ParticleSampleable{SingleParticleVariate})
+    ```
+    which returns a random sample from `s` as a `ParticleStateful`.
+
+!!! note "Multiple particle sampler"
+    For `MultiParticleVariate` samplers, one needs to implement
+
+    ```julia
+    Base.size(s::ParticleSampleable{MultiParticleVariate})
+    ```
+    and the mutating version of `rand`:
+
+    ```julia
+    Distributions._rand!(
+        rng::Random.AbstractRNG,
+        s::ParticleSampleable{MultiParticleVariate},
+        out::AbstractArray{ParticleStateful})
+    ```
+
+!!! note "Scattering process sampler"
+    For `ProcessLikeVariate` samplers, one needs to give the single sampler version of `rand`:
+
+    ```julia
+    Distributions.rand(
+        rng::Random.AbstractRNG,
+        s::ParticleSampleable{SingleParticleVariate})
+    ```
+    which must a random sample from `s` as a `PhaseSpacePoint` of the respective scattering process.
+
+"""
+abstract type ParticleSampleable{F<:QEDlikeVariate} <:
+              Distributions.Sampleable{F,Distributions.Continuous} end
+
+"""
+    _momentum_type(s::ParticleSampleable,x)
+
+Return the momentum type used for the generation of samples. The default is `SFourMomentum`.
+
+!!! note
+
+    This interface function is optional for subtypes of [`ParticleSampleable`](@ref).
+
+"""
+function _momentum_type(s::ParticleSampleable)
+    return SFourMomentum
+end
+
+"""
+    _assert_valid_input_type(s::ParticleSampleable,x)
+
+
+Throw `InvalidInputError` if the input `x` has the wrong type, do nothing otherwise. This function is usually used for complicated types, where
+the implementation via multiple dispatch is cumbersome. The default is doing nothing.
+
+!!! note
+
+    This interface function is optional for subtypes of [`ParticleSampleable`](@ref).
+
+"""
+function _assert_valid_input_type(s::ParticleSampleable, x) end
```

Comment:
Actually, we have all the information to check the type match between the `ParticleSampleable` and the input `x` for some sub-interfaces, e.g. `SingleParticleDistribution`. Therefore, it makes sense to provide defaults, instead of leaving them on the side of the custom implementation. So we need to distinguish between type checking and general validity checking in the first place.

BTW: one could do similar things for the size checks as well.